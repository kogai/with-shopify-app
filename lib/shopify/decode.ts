import { NextApiRequest, GetServerSidePropsContext } from 'next'
import querystring from 'querystring'
import { VerifyPayload } from './client'
import { createErr, createOk, Result } from '../result'

export const decodeVerifyPayload = ({
  hmac,
  code,
  shop,
  state,
  timestamp,
  ...rest
}: NextApiRequest['query']): Result<VerifyPayload, Error> => {
  if (
    typeof hmac !== 'string' ||
    typeof code !== 'string' ||
    typeof shop !== 'string' ||
    typeof state !== 'string' ||
    typeof timestamp !== 'string' ||
    Object.values(rest).some((x) => typeof x !== 'string')
  ) {
    return createErr(
      new Error(
        `invalid property, ${querystring.stringify({
          hmac,
          code,
          shop,
          state,
          timestamp,
          ...rest,
        })}`
      )
    )
  }

  return createOk({
    hmac,
    code,
    shop,
    state,
    timestamp,
    ...rest,
  })
}

type ParsedUrlQuery = GetServerSidePropsContext['query']

type IndexQuery =
  | {
      type: 'pre'
      shop: string
    }
  | {
      type: 'post'
      accessToken: string
      storeFrontAccessToken: string
      shop: string
    }

export const decodeIndexQuery = (
  query: ParsedUrlQuery
): Result<IndexQuery, Error> => {
  const { shop, accessToken, storeFrontAccessToken } = query
  if (typeof shop !== 'string') {
    return createErr(new Error(`Invalid access. shop=${shop}`))
  }
  if (
    typeof accessToken === 'string' &&
    typeof storeFrontAccessToken === 'string'
  ) {
    return createOk({
      type: 'post',
      shop,
      storeFrontAccessToken,
      accessToken,
    })
  }
  return createOk({
    type: 'pre',
    shop,
  })
}
