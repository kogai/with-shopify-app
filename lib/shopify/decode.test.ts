import { decodeVerifyPayload, decodeIndexQuery } from './decode'

describe('lib/shopify/decode', () => {
  describe('decodeVerifyPayload', () => {
    const standardPayload = {
      hmac: 'abc',
      code: 'def',
      shop: 'ghi',
      state: 'jkl',
      timestamp: 'mno',
      a: 'pqr',
    }

    ;[
      {
        title: 'ok',
        input: {
          ...standardPayload,
        },
      },
      {
        title: 'ng(bound property)',
        input: {
          ...standardPayload,
          code: ['ABC', 'DEF'],
        },
      },
      {
        title: 'ng(free property)',
        input: {
          ...standardPayload,
          a: ['ABC', 'DEF'],
        },
      },
    ].forEach(({ title, input }) => {
      it(title, () => {
        expect(decodeVerifyPayload(input)).toMatchSnapshot()
      })
    })
  })

  describe('decodeIndexQuery', () => {
    ;[
      {
        title: 'ok(pre)',
        input: {
          shop: 'mytest.myshopify.com',
        },
      },
      {
        title: 'ok(post)',
        input: {
          shop: 'mytest.myshopify.com',
          accessToken: 'abc',
          storeFrontAccessToken: 'def',
        },
      },
      {
        title: 'ng',
        input: {},
      },
      {
        title: 'ng(bound property)',
        input: {
          shop: ['mytest1.myshopify.com', 'mytest2.myshopify.com'],
        },
      },
    ].forEach(({ title, input }) => {
      it(title, () => {
        expect(decodeIndexQuery(input)).toMatchSnapshot()
      })
    })
  })
})
