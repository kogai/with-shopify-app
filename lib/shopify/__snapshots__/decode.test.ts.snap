// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/shopify/decode decodeIndexQuery ng 1`] = `
Object {
  "err": [Error: Invalid access. shop=undefined],
  "ok": false,
  "val": undefined,
}
`;

exports[`lib/shopify/decode decodeIndexQuery ng(bound property) 1`] = `
Object {
  "err": [Error: Invalid access. shop=mytest1.myshopify.com,mytest2.myshopify.com],
  "ok": false,
  "val": undefined,
}
`;

exports[`lib/shopify/decode decodeIndexQuery ok(post) 1`] = `
Object {
  "err": undefined,
  "ok": true,
  "val": Object {
    "accessToken": "abc",
    "shop": "mytest.myshopify.com",
    "storeFrontAccessToken": "def",
    "type": "post",
  },
}
`;

exports[`lib/shopify/decode decodeIndexQuery ok(pre) 1`] = `
Object {
  "err": undefined,
  "ok": true,
  "val": Object {
    "shop": "mytest.myshopify.com",
    "type": "pre",
  },
}
`;

exports[`lib/shopify/decode decodeVerifyPayload ng(bound property) 1`] = `
Object {
  "err": [Error: invalid property, hmac=abc&code=ABC&code=DEF&shop=ghi&state=jkl&timestamp=mno&a=pqr],
  "ok": false,
  "val": undefined,
}
`;

exports[`lib/shopify/decode decodeVerifyPayload ng(free property) 1`] = `
Object {
  "err": [Error: invalid property, hmac=abc&code=def&shop=ghi&state=jkl&timestamp=mno&a=ABC&a=DEF],
  "ok": false,
  "val": undefined,
}
`;

exports[`lib/shopify/decode decodeVerifyPayload ok 1`] = `
Object {
  "err": undefined,
  "ok": true,
  "val": Object {
    "a": "pqr",
    "code": "def",
    "hmac": "abc",
    "shop": "ghi",
    "state": "jkl",
    "timestamp": "mno",
  },
}
`;
